### PR TITLE
core: Implement HasAncestor.

### PR DESCRIPTION
**Pull request by PapyChacal**
Implement HasAncestor, the transitive cousin of HasParent.

### CODE REVIEW

File: xdsl/traits.py [hunk 1]
```diff
@@ -68,6 +68,40 @@ def verify(self, op: Operation) -> None:
...
+    Constraint the operation to have a specific operation as ancestor, i.e. transitive
+    parent.
+    """
+
+    parameters: tuple[type[Operation], ...]
+
+    def __init__(self, *parameters: type[Operation]):
+        if not parameters:
+            raise ValueError("parameters must not be empty")
+        super().__init__(parameters)
```
Comment (superlopuh) [inline xdsl/traits.py:83]: One way to enforce this syntactically is to have a first param, and remaining params
```suggestion
    def __init__(self, head_param: type[Operation], *tail_parameters: type[Operation]):
        super().__init__((head_param, *parameters))
```
Seems like the tuple construction cost is minimal compared to the runtime error annoyance, but that's kind of a matter of taste :)

File: xdsl/traits.py [hunk 2]
```diff
@@ -68,6 +68,40 @@ def verify(self, op: Operation) -> None:
...
+                f"'{op.name}' expects ancestor op to be one of {names}"
+            )
+
+    def get_ancestor(self, op: Operation) -> Operation | None:
+        parent = op.parent_op()
+        while not isinstance(parent, self.parameters):
+            if parent is None:
+                return None
+            parent = parent.parent_op()
+        return parent
```
Comment (superlopuh) [inline xdsl/traits.py:102]: Should this be a helper on operation? It has a very similar vibe to `walk` and `walk_backwards`, I think a `walk_ancestors` would make sense. To minimise the scope creep, I'd like to propose splitting the iteration from the predicate first, putting both helpers here, and then eventually migrate them.
```suggestion
    def walk_ancestors(self, op: Operation) -> Iterator[Operation]:
        """Iterates over the ancestors of an operation, including the input"""
        curr = op
        yield curr
        while (curr := curr.parent_op()) is not None:
            yield curr

    def get_ancestor(self, op: Operation) -> Operation | None:
        ancestors = self.walk_ancestors(op)
        matching_ancestors = (a for a in ancestors if isinstance(a, self.parameters))
        return next(matching_ancestors, None)
```

File: xdsl/traits.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/traits.py:83]: Right, will do 🙂

File: xdsl/traits.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/traits.py:102]: Why not!

Comment (PapyChacal) [reply xdsl/traits.py:100]: Will do but if anyone else disagrees and we start bikeshedding on core, I'm reversing to add functionality first :+1:

Comment (PapyChacal) [reply xdsl/traits.py:100]: Oh wait; didn't see your idea of splitting but still in traits. Perfect on the concern ahah

**Review by superlopuh — APPROVED** [review]
